File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
from collections import deque
for _ in range(int(input())):
    N = int(input())

    end = list(map(int, input().split()))
    visit = [False for _ in range(N)]
    ans = 0
    for i in range(N):
        if not visit[i]:
            ans += 1
            queue = deque()
            queue.append(i)
            visit[i] = True

            while queue:
                x = queue.popleft()
                if not visit[end[x]-1]:
                    visit[end[x]-1] = True
                    queue.append(end[x]-1)

    print(ans)

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {

	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 
        int T = Integer.parseInt(br.readLine());

        for (int t = 0; t < T; t++) {
            int N = Integer.parseInt(br.readLine());
            ArrayList<Integer> end = new ArrayList<>();
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < N; i++) {
                end.add(Integer.parseInt(st.nextToken()));
            }

            boolean[] visit = new boolean[N];
            int ans = 0;

            for (int i = 0; i < N; i++) {
                if (!visit[i]) {
                    ans++;
                    Queue<Integer> queue = new LinkedList<>();
                    queue.add(i);
                    visit[i] = true;

                    while (!queue.isEmpty()) {
                        int x = queue.poll();
                        int nextIndex = end.get(x) - 1;
                        if (!visit[nextIndex]) {
                            visit[nextIndex] = true;
                            queue.add(nextIndex);
                        }
                    }
                }
            }
            System.out.println(ans);
        }
	}
}